ex5/h_miyaji/main.py

```diff
@@ -0,0 +1,395 @@
+    parser.add_argument("--input-file", type=str, required=True, help="input csv file")
+
+    # number of clusters for Gaussian distribution
+    parser.add_argument("--cluster-num", type=int, default=1, help="number of clusters")
+
+    # convergence condition for EM algorithm
+    parser.add_argument(
+        "--error", type=float, default=0.001, help="convergence condition"
+    )
+
+    return parser.parse_args()
+
+
+def initialize_gmm(
+    dim: int, cluster_num: int
+) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Initialize the means, covariance matrix and weights.
+
+    Args:
+        dim (int): the dimension of csv data.
+        cluster_num (int): number of clusters.
+
+    Returns:
+        means (np.ndarray): mean for each cluster (k, dim).
+        cov_matrix (np.ndarray): covariance matrix for each cluster (k, dim, dim).
+        weights (np.ndarray): weight of each Gaussian (k, ).
+    """
+    means = np.random.randn(cluster_num, dim)  # (k, dim)
+    cov_matrix = np.array([np.eye(dim) for k in range(cluster_num)])  # (k, dim, dim)
+    weights = np.array([1 / cluster_num for k in range(cluster_num)])  # (k, )
+    return means, cov_matrix, weights
+
+
+def get_gauss(data: np.ndarray, mean: np.ndarray, cov_matrix: np.ndarray) -> float:
+    """Get the value of the Gaussian distribution.
+
+    Args:
+        data (np.ndarray): one of the csv data (dim,).
+        mean (np.ndarray): mean in one cluster (dim,).
+        cov_matrix (np.ndarray): covariance matrix in one cluster (dim, dim).
+
+    Returns:
+        gauss (float): the value of the Gaussian distribution.
+    """
+    # dimension of the data
+    if isinstance(data, np.ndarray):
+        dim = len(data)
+    else:
+        dim = 1
+
+    gauss_denominator = (np.sqrt(2 * np.pi) ** dim) * np.sqrt(np.linalg.det(cov_matrix))
+    gauss_numerator = np.exp(
+        ((data - mean) @ np.linalg.inv(cov_matrix) @ (data - mean)[:, np.newaxis]) / -2
+    )
+
+    gauss = gauss_numerator.item() / gauss_denominator.item()
+    return gauss
+
+
+def calc_mix_gauss(
+    dataset: np.ndarray,
+    means: np.ndarray,
+    cov_matrix: np.ndarray,
+    weights: np.ndarray,
+) -> np.ndarray:
+    """Calculate values of mix Gaussian distribution.
+
+    Args:
+        dataset (np.ndarray): csv data (n, dim).
+        means (np.ndarray): mean for each cluster (k, dim).
+        cov_matrix (np.ndarray): covariance matrix for each cluster (k, dim, dim).
+        weights (np.ndarray): weight of each Gaussian (k, ).
+
+    Returns:
+        mix_gauss (np.ndarray): values of mix Gaussian distribution (k, n).
+    """
+    cluster_num = len(weights)  # number of Gaussian
+    data_num = len(dataset)  # number of csv data
+
+    mix_gauss = np.zeros((cluster_num, data_num))  # (k, n)
+
+    for k in range(cluster_num):
+        for n in range(data_num):
+            mix_gauss[k, n] = weights.item(k) * get_gauss(
+                dataset[n], means[k], cov_matrix[k]
+            )
```

Comment:
> 97, 139, 150行目付近のforループを行列計算にできれば良かったのですが、方法が分からなかったため、そのままにしてあります。いい方法があれば教えていただきたいです。

考えてみましたが，こんな感じでforループをなくせると思います．ただ，クラスター数の方向にmapを使っているため割と屁理屈っぽいですが… もしかしたらmapも内包表記も無しで実装できるかもです．
```suggestion
    from scipy.stats import multivariate_normal
    
    gaussian_pdfs = list(
        map(lambda mean, cov: multivariate_normal(mean, cov), means, cov_matrix)
    )
    mix_gauss = np.array(
        list(
            map(
                lambda pdf, weight: pdf.pdf(dataset) * weight,
                gaussian_pdfs,
                weights,
            )
        )
    )
```